### PR TITLE
feat!: add shims config

### PR DESCRIPTION
**Pull request by fi3ework**
## Summary

💥 Breaking changes:

- `__dirname` and `__filename` won't be shimmed in ESM output by default. Set `lib.shims.esm.__dirname` and `lib.shims.esm.__filename` to `true` to enable.

## Checklist

- [x] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: packages/core/src/plugins/shims.ts [hunk 1]
```diff
@@ -11,9 +11,8 @@ const importMetaUrlShim = `/*#__PURE__*/ (function () {
 // - Replace `import.meta.url` with `importMetaUrl`.
 // - Inject `importMetaUrl` to the end of the module (can't inject at the beginning because of `"use strict";`).
 // This is a short-term solution, and we hope to provide built-in polyfills like `node.__filename` on Rspack side.
-export const pluginCjsShim = (): RsbuildPlugin => ({
+export const pluginCjsImportMetaUrlShim = (): RsbuildPlugin => ({
   name: 'rsbuild-plugin-cjs-shim',
```
Comment (Timeless0911) [inline packages/core/src/plugins/shims.ts:15]: better to modify the name too

**Comment by Timeless0911** [comment]
The PR title can be a feature about adding shims config.

File: packages/core/src/config.ts [hunk 2]
```diff
@@ -531,6 +542,85 @@ const composeFormatConfig = (format: Format): RsbuildConfig => {
...
+    return resolvedShims;
+  }
+
+  if (shims.cjs) {
+    if (typeof shims.cjs === 'boolean') {
+      if (shims.cjs === true) {
+        resolvedShims.cjs['import.meta.url'] = true;
+      } else {
+        resolvedShims.cjs['import.meta.url'] = false;
+      }
```
Comment (Timeless0911) [inline packages/core/src/config.ts:567]: only need to deal when `shims.cjs === false`? `resolvedShims.cjs['import.meta.url']` default to true

File: packages/core/src/config.ts [hunk 3]
```diff
@@ -531,6 +542,85 @@ const composeFormatConfig = (format: Format): RsbuildConfig => {
   }
 };
 
+const resolveShims = (shims?: Shims) => {
+  const resolvedShims = {
```
Comment (Timeless0911) [inline packages/core/src/config.ts:546]: default value TBD

File: packages/core/src/config.ts [hunk 4]
```diff
@@ -531,6 +542,52 @@ const composeFormatConfig = (format: Format): RsbuildConfig => {
   }
 };
 
+const resolveShims = (shims?: Shims): ResolvedShims => {
```
Comment (Timeless0911) [inline packages/core/src/config.ts:545]: why not write directly into `composeShimsConfig`

File: packages/core/src/plugins/shims.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    : (document.currentScript && document.currentScript.src) ||
+      new URL('main.js', document.baseURI).href;
+})()`;
+
+// This Rsbuild plugin will shim `import.meta.url` for CommonJS modules.
+// - Replace `import.meta.url` with `importMetaUrl`.
+// - Inject `importMetaUrl` to the end of the module (can't inject at the beginning because of `"use strict";`).
+// This is a short-term solution, and we hope to provide built-in polyfills like `node.__filename` on Rspack side.
+export const pluginCjsImportMetaUrlShim = (): RsbuildPlugin => ({
+  name: 'rsbuild-plugin-cjs-import-meta-url-shim',
```
Comment (Timeless0911) [inline packages/core/src/plugins/shims.ts:15]:
```suggestion
  name: 'rsbuild:cjs-import-meta-url-shim',
```

File: packages/core/src/plugins/shims.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  },
+});
+
+const requireShim = `// Rslib ESM shims
+import __rslib_shim_module__ from 'module';
+const require = /*#__PURE__*/ __rslib_shim_module__.createRequire(import.meta.url);
+`;
+
+export const pluginEsmRequireShim = (): RsbuildPlugin => ({
+  name: 'rsbuild-plugin-esm-require-shim',
```
Comment (Timeless0911) [inline packages/core/src/plugins/shims.ts:32]:
```suggestion
  name: 'rsbuild:esm-require-shim',
```

File: packages/core/src/types/config/index.ts [hunk 7]
```diff
@@ -48,6 +48,22 @@ export type BannerAndFooter = {
...
+    'import.meta.url'?: boolean;
+  };
+  esm?: {
+    __filename?: boolean;
+    __dirname?: boolean;
+    require?: boolean;
+  };
+};
+
+export type ResolvedShims = {
```
Comment (Timeless0911) [inline packages/core/src/types/config/index.ts:62]: should be updated?

File: packages/core/tests/__snapshots__/config.test.ts.snap [hunk 8]
```diff
@@ -196,7 +201,7 @@ exports[`Should compose create Rsbuild config correctly > Merge Rsbuild config 1
       },
       "plugins": [
         {
-          "name": "rsbuild-plugin-cjs-shim",
+          "name": "rsbuild-plugin-cjs-import-meta-url-shim",
```
Comment (Timeless0911) [inline packages/core/tests/__snapshots__/config.test.ts.snap:204]:
```suggestion
          "name": "rsbuild:cjs-import-meta-url-shim",
```

File: packages/core/src/types/config/index.ts (hunk 7 above)
Comment (fi3ework) [reply packages/core/src/types/config/index.ts]: Removed the type directly.

**Review by Timeless0911 — APPROVED** [review]
LGTM

**Pull request merged**